### PR TITLE
Fixes agent merging failing unexpectedly

### LINKED ISSUE #4685: Agent merge failing unexpectedly
**Describe the bug**

**To Reproduce**

1. Go to https://herb32124-edge.test.specifysystems.org/specify/overlay/merge/Agent/?records=299,11910
2. Merge the two agents
3. See error

**Error:**

```json
{
   "message": "Traceback (most recent call last):
   File \"/opt/specify7/specifyweb/specify/record_merging.py\", line 47, in resolve_record_merge_response
   response = start_function()
   File \"/opt/specify7/specifyweb/specify/record_merging.py\", line 389, in <lambda>
   lambda: record_merge_fx(model_name, old_model_ids, int(new_model_id), progress, new_record_info))
   File \"/usr/lib/python3.8/contextlib.py\", line 75, in inner
   return func(*args, **kwds)
   File \"/opt/specify7/specifyweb/specify/record_merging.py\", line 315, in record_merge_fx
   target_model.objects.get(id=old_model_id).delete()
   File \"/opt/specify7/ve/lib/python3.8/site-packages/django/db/models/base.py\", line 966, in delete
   collector.collect([self], keep_parents=keep_parents)
   File \"/opt/specify7/ve/lib/python3.8/site-packages/django/db/models/deletion.py\", line 302, in collect
   raise ProtectedError(
   django.db.models.deletion.ProtectedError: (\"Cannot delete some instances of model 'Agent' because they are referenced through protected foreign keys: 'Collectionobject.cataloger'.\", {<Collectionobject: Collectionobject object (200994)>, <Collectionobject: Collectionobject object (200995)>, <Collectionobject: Collectionobject object (201000)>, <Collectionobject: Collectionobject object (201001)>, <Collectionobject: Collectionobject object (201002)>, <Collectionobject: Collectionobject object (201003)>, <Collectionobject: Collectionobject object (201004)>, <Collectionobject: Collectionobject object (201005)>, <Collectionobject: Collectionobject object (201006)>, <Collectionobject: Collectionobject object (201007)>, <Collectionobject: Collectionobject object (201008)>, <Collectionobject: Collectionobject object (201009)>, <Collectionobject: Collectionobject object (201010)>, <Collectionobject: Collectionobject object (201011)>, <Collectionobject: Collectionobject object (201012)>, <Collectionobject: Collectionobject object (201013)>, <Collectionobject: Collectionobject object (201014)>, <Collectionobject: Collectionobject object (201015)>, <Collectionobject: Collectionobject object (201016)>, <Collectionobject: Collectionobject object (201017)>, <Collectionobject: Collectionobject object (201018)>, <Collectionobject: Collectionobject object (201019)>, <Collectionobject: Collectionobject object (201020)>})"
}

```

**Screenshots**

https://herb32124-edge.test.specifysystems.org/specify/overlay/merge/Agent/?records=299,11910

https://github.com/specify/specify7/assets/37256050/77f21012-07b5-465f-bd5c-76ebd53ffbb7

**Crash Report**
[Michael Todd – Merging 6282219c-e2d1-4943-9b44-aac1173469ea Crash Report - 2024-03-25T19_03_12.668Z.txt](https://github.com/specify/specify7/files/14748366/Michael.Todd.Merging.6282219c-e2d1-4943-9b44-aac1173469ea.Crash.Report.-.2024-03-25T19_03_12.668Z.txt)

**Reported by:** Royal Botanic Gardens Edinburgh

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #4685

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add relevant issue to release milestone

### Testing instructions

- use herb32124 db
- merge two agents 

- [ ] verify it works 

- merge any other agents with any other db and verify it works as expected

### CODE REVIEW

File: specifyweb/specify/record_merging.py [hunk 1]
```diff
@@ -308,7 +308,7 @@ def update_record(record: models.Model):
 
                 response: http.HttpResponse = update_record(obj)
                 if response is not None and response.status_code != 204:
```
Comment (realVinayak) [inline specifyweb/specify/record_merging.py:310]: @CarolineDenis thanks for fixing this!

@acwhite211 on second thought, do we even need this now? I added handling the invalid http responses through FailedMergingException to preserve atomicity, which gets caught by the caller of this function. See [here](https://github.com/specify/specify7/blob/075104652b2204e273501289a7090e79b3998007/specifyweb/specify/record_merging.py#L45-L64).


So, we can get rid of all the HTTP responses [here]. (https://github.com/specify/specify7/blob/075104652b2204e273501289a7090e79b3998007/specifyweb/specify/record_merging.py#L150-L350). So now, this function won't return anything but would raise a FailedMergingException or any other exception. 

Here, now we define the 204 http in the try block, and return that.

https://github.com/specify/specify7/blob/075104652b2204e273501289a7090e79b3998007/specifyweb/specify/record_merging.py#L45-L64

**Comment by grantfitzsimmons** [comment]
> ### Testing instructions
> 
> - use herb32124 db
> - merge two agents 
> 
> - [ ] verify it works 

I believe the testing instructions for record merging should be more detailed to ensure it functions correctly in various scenarios, not just this one.

File: specifyweb/specify/record_merging.py [hunk 2]
```diff
@@ -197,13 +197,13 @@ def record_merge_fx(model_name: str, old_model_ids: List[int], new_model_id: int
...
             table_name.lower() in MERGING_OPTIMIZATION_FIELDS[model_name.lower()]:
-            for field_name in MERGING_OPTIMIZATION_FIELDS[model_name.lower()][table_name.lower()]:
+            if field_name in MERGING_OPTIMIZATION_FIELDS[model_name.lower()][table_name.lower()]:
                 query = Q(**{field_name: old_model_ids[0]})
                 for old_model_id in old_model_ids[1:]:
                     query.add(Q(**{field_name: old_model_id}), Q.OR)
                 foreign_model.objects.filter(query).update(**{field_name: new_model_id})
                 progress(1, 0) if progress is not None else None
-            continue
+                continue
```
Comment (maxpatiiuk) [inline specifyweb/specify/record_merging.py:206]: `continue` is useless here as you are not in a `for` anymore

Comment (CarolineDenis) [reply specifyweb/specify/record_merging.py:206]: ha ok, I added it back under @realVinayak advise

Comment (realVinayak) [reply specifyweb/specify/record_merging.py:206]: because there is a for loop a couple of lines up!

**Review by maxpatiiuk — APPROVED** [review]

**Comment by melton-jason** [comment]
@specify/ux-testing 
The issue encountered in https://github.com/specify/specify7/issues/4685#issue-2206501707 can be recreated by merging Agents in which at least two Agents are referenced by a `CollectionObject -> Cataloger` relationship.

To recreate the Issue from scratch: 
 - Create two Agents
 - Assign each as a Cataloger to two separate CollectionObjects
 - Attempt to merge the Agents



In the below video, both Agents are referenced by at least one `CollectionObject -> Cataloger` (as seen through the Linked Records display) 

https://github.com/specify/specify7/assets/64045831/3ae5ce47-8a16-4866-ba81-2a49aa2cc45d

Other Agents can be associated with the merge (and even made as the 'Primary' agent) and the merge will still fail regardless of the Linked records. 


https://github.com/specify/specify7/assets/64045831/0e28911e-5327-4e6b-8e83-03cf7b7d5bef

**Review by alesan99 — APPROVED** [review]
- use herb32124 db
- merge two agents 
- [X] verify it works 
- merge any other agents with any other db and verify it works as expected 

I recreated the issue on KUFish by merging two agents linked to separate Collection Objects.
Merging now works successfully in this case 👍👍

https://github.com/specify/specify7/assets/8108710/2d23f9e1-5ddc-45b4-8c5a-fbe4f50b5bf6

I also tested it with merging separate gift agents and unlinked agents. I could not get these cases to cause an error on production though.

**Pull request merged**